org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/ServiceUseRuleTest.java

```diff
@@ -0,0 +1,317 @@
+import org.junit.Test;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.test.common.osgi.ServiceUse;
+import org.osgi.test.junit4.types.Foo;
+
+public class ServiceUseRuleTest {
+
+	@Test
+	public void basicAssumptions() throws Exception {
+		try (BundleContextRule bcRule = new BundleContextRule();
+			ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule).cardinality(0)
+				.build()) {
+
+			bcRule.init(getClass());
+			foos.init(getClass());
+
+			SoftAssertions softly = new SoftAssertions();
+
+			softly.assertThat(foos.isEmpty())
+				.isTrue();
+			softly.assertThat(foos.getTimeout())
+				.isEqualTo(ServiceUse.DEFAULT_TIMEOUT);
+			softly.assertThat(foos.getCardinality())
+				.isEqualTo(0);
+
+			softly.assertAll();
+		}
+	}
+
+	@Test
+	public void requiredFailsWhenNoService() throws Exception {
+		Throwable throwable = catchThrowable(() -> {
+			try (BundleContextRule bcRule = new BundleContextRule();
+				ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule).cardinality(1)
+					.build()) {
+
+				bcRule.init(getClass());
+				foos.init(getClass());
+			}
+		});
+
+		assertThat(throwable).isInstanceOf(AssertionError.class)
+			.hasMessageContaining(
+				" services (objectClass=org.osgi.test.junit4.types.Foo) didn't arrive within 200ms");
+	}
+
+	@Test
+	public void requiredFailsWhenNoServiceWithTimeout() throws Exception {
+		Throwable throwable = catchThrowable(() -> {
+			try (BundleContextRule bcRule = new BundleContextRule();
+				ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule).timeout(10)
+					.build()) {
+
+				bcRule.init(getClass());
+				foos.init(getClass());
+			}
+		});
+
+		assertThat(throwable).isInstanceOf(AssertionError.class)
+			.hasMessageContaining(
+				" services (objectClass=org.osgi.test.junit4.types.Foo) didn't arrive within 10ms");
+	}
+
+	@Test
+	public void successWhenService() throws Exception {
+		try (BundleContextRule bcRule = new BundleContextRule();
+			ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule).cardinality(1)
+				.build();) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, null),
+					100, TimeUnit.MILLISECONDS);
+
+			foos.init(getClass());
+
+			SoftAssertions softly = new SoftAssertions();
+
+			softly.assertThat(foos.isEmpty())
+				.isFalse();
+			softly.assertThat(foos.size())
+				.isEqualTo(1);
+			softly.assertThat(foos.getTimeout())
+				.isEqualTo(ServiceUse.DEFAULT_TIMEOUT);
+			softly.assertThat(foos.getCardinality())
+				.isGreaterThan(0);
+
+			softly.assertAll();
+		}
+	}
+
+	@Test
+	public void successWhenServiceWithTimeout() throws Exception {
+		try (BundleContextRule bcRule = new BundleContextRule();
+			ServiceUseRule<Foo> foos = new ServiceUseRule.Builder<>(Foo.class, bcRule)
+				.timeout(10)
```

Comment:
This test effectively has only a 5ms tolerance - if for any reason your the scheduled service registration is delayed by more than 5ms, or takes more than 5ms to complete, then the test will fail. I think that 5ms tolerance is perhaps a bit ambitious. I'd make this timeout 100ms - this only slows down the test in the case of a test failure and makes the test less prone to flakiness.

Reply:
fixing...